qmi/instruments/thorlabs/apt_protocol.py

```diff
@@ -0,0 +1,249 @@
+    HW_START_UPDATEMSGS = 0x0011
+    HW_STOP_UPDATEMSGS = 0x0012
+    MOT_MOVE_HOME = 0x0443
+    MOT_MOVE_HOMED = 0x0444
+    MOT_MOVE_ABSOLUTE = 0x0453
+    MOT_MOVE_COMPLETED = 0x0464
+    MOT_SET_EEPROMPARAMS = 0x04B9
+    MOT_REQ_USTATUSUPDATE = 0x0490
+    MOT_GET_USTATUSUPDATE = 0x0491
+    MOT_MOVE_JOG = 0x046A
+    POL_SET_PARAMS = 0x0530
+    POL_REQ_PARAMS = 0x0531
+    POL_GET_PARAMS = 0x0532
+
+
+class AptChannelState(Enum):
+    """Channel state"""
+
+    ENABLE = 0x01
+    DISABLE = 0x02
+
+
+class AptChannelJogDirection(Enum):
+    """Jog direction"""
+
+    FORWARD = 0x01
+    BACKWARD = 0x02
+
+
+class AptMessageHeaderWithParams(LittleEndianStructure):
+    """
+    This is the version of the APT message header when no data packet follows a header.
+    """
+
+    _pack_ = True
+    _fields_: List[Tuple[str, type]] = [
+        ("message_id", c_uint16),
+        ("param1", c_uint8),
+        ("param2", c_uint8),
+        ("dest", c_uint8),
+        ("source", c_uint8),
+    ]
+
+
+class AptMessageHeaderForData(LittleEndianStructure):
+    """
+    This is the version of the APT message header when a data packet follows a header.
+    """
+
+    _pack_ = True
+    _fields_: List[Tuple[str, type]] = [
+        ("message_id", c_uint16),
+        ("date_length", c_uint16),
+        ("dest", c_uint8),
+        ("source", c_uint8),
+    ]
+
+
+class AptMessage(LittleEndianStructure):
+    """
+    Base class for an APT message.
+    """
+
+    MESSAGE_ID: int
+    HEADER_ONLY: bool = False
+    _pack_ = True
+
+
+T = TypeVar("T", bound=AptMessage)
+
+
+class AptProtocol:
+    """
+    Implement the Thorlabs APT protocol primitives.
+    """
+
+    HEADER_SIZE_BYTES = 6
+
+    def __init__(
+        self,
+        transport: QMI_Transport,
+        apt_device_address: int = 0x50,
+        host_address: int = 0x01,
+        default_timeout: Optional[float] = None,
+    ):
+        """Initialize the Thorlabs APT protocol handler.
+
+        Parameters:
+            transport:          Instance of `QMI_Transport` to use for sending APT commands to the instrument.
+            apt_device_address: The address of the APT device. By default it is 0x50 which is a generic USB hardware
+                                unit.
+            host_address:       The address of the host that sends and receives messages. By default it is 0x01 which
+                                is a host controller such as a PC.
+            default_timeout:    Optional default response timeout in seconds.
+                                The default is to wait indefinitely until a response is received.
+        """
+        self._transport = transport
+        self._timeout = default_timeout
+        self._apt_device_address = apt_device_address
+        self._host_address = host_address
+
+    def write_param_command(
+        self,
+        message_id: int,
+        param1: Optional[int] = None,
+        param2: Optional[int] = None,
```

Comment:
Couldn't you just do `int = 0`? Then in the message you just use `param1`, `param2` as 0x00 == 0.

Reply:
yes! good catch